.mage/js.go

```diff
@@ -185,6 +184,28 @@ func (js Js) BuildMain() error {
 	return webpack("--config", "config/webpack.config.babel.js")
 }
 
+// BuildIfNecessary builds the frontend bundles but only if source files have
+// been modified.
+func (js Js) BuildIfNecessary() error {
+	isCI := os.Getenv("CI") == "true"
+	publicExists := pathExists("./public")
+	changed, err := target.Dir("./public", "./pkg/webui", "./config/webpack.config.babel.js", "./config/webpack.dll.babel.js", "./sdk/js", "./package.json")
+	if err != nil {
+		return err
+	}
+	if (changed && !isCI) || !publicExists {
+		if mg.Verbose() {
+			fmt.Println("Building frontend files due to changes in source files...")
+		}
+		mg.Deps(js.Build)
+	} else {
+		if mg.Verbose() {
+			fmt.Println("Skipping frontend build due to no source file changes being detected...")
+		}
+	}
+	return nil
+}
+
```

Comment:
`js:build` *is* `js:buildIfNecessary` - if it's not, it's a bug. Remove
```suggestion
```

Reply:
From my experience, it's not desirable to apply target checks for the `js.build` as they will always be quite fragile, and result in annoying false positives too often. I really want this command to always execute `js.buildMain`, which is why that target did not check for anything and would always execute, sinxe it's not possible to force execution regardless of targets, as it was in `make`. 

For the CI however, it's good to add these checks to avoid unnecessary builds when the caches are available.

Reply:
Refactored this to include the CI check in the normal `Js.Build` target.